relentless/potential/pair.py

```diff
@@ -156,6 +155,270 @@ def _derivative(self, param, r, epsilon, sigma, **params):
+        if self.mode == 'diff':
+            # difference is next knot minus my knot, with last knot fixed at its current value
+            ks[:-1] -= ks[1:]
+
+        # make all knots variables, but hold all r and the last knot const
+        for i in range(self.num_knots):
+            ri,ki = self._knot_params(i)
+            if self.coeff[pair][ri] is None:
+                self.coeff[pair][ri] = core.Variable(rs[i],const=True)
+            else:
+                self.coeff[pair][ri].value = rs[i]
+            if self.coeff[pair][ki] is None:
+                self.coeff[pair][ki] = core.Variable(ks[i],const=(i==self.num_knots-1))
+            else:
+                self.coeff[pair][ki].value = ks[i]
+
+    def _knot_params(self, i):
+        """Get the parameter names for a given knot.
+
+        Parameters
+        ----------
+        i : int
+            Key for the knot variable.
+
+        Returns
+        -------
+        str
+            The keyed r value.
+        str
+            The keyed knot value.
+
+        Raises
+        ------
+        ValueError
+            If the knot key is not an integer.
+
+        """
+        if not isinstance(i, int):
+            raise ValueError('Knots are keyed by integers')
+        return 'r-{}'.format(i), 'knot-{}'.format(i)
+
+    def _energy(self, r, **params):
+        """Evaluates the spline potential energy.
+
+        Parameters
+        ----------
+        r : array_like
+            The value or values of r at which to evaluate the energy.
+        params : dict
+            The parameters (r and knot values) for the potential function.
+
+        Returns
+        -------
+        array_like
+            Returns the value or values of the energy evaluated at r.
+
+        """
+        r,u,s = self._zeros(r)
+        u = self._interpolate(params)(r)
+        if s:
+            u = u.item()
+        return u
+
+    def _force(self, r, **params):
+        """Evaluates the spline force.
+
+        Parameters
+        ----------
+        r : array_like
+            The value or values of r at which to evaluate the energy.
+        params : dict
+            The parameters (r and knot values) for the potential function.
+
+        Returns
+        -------
+        array_like
+            Returns the value or values of the force evaluated at r.
+
+        """
+        r,f,s = self._zeros(r)
+        f = -self._interpolate(params).derivative(r, 1)
+        if s:
+            f = f.item()
+        return f
+
+    def derivative(self, pair, param, r):
+        #Extending PairPotential method to check if r and knot values are Variables.
+        for i in range(self.num_knots):
+            ri,ki = self._knot_params(i)
+            if not isinstance(self.coeff[pair][ri], core.Variable):
+                raise TypeError('All r values must be Variables')
+            if not isinstance(self.coeff[pair][ki], core.Variable):
+                raise TypeError('All knot values must be Variables')
```

Comment:
Maybe simplify with the knot generator?
```suggestion
        for ri,ki in self.knots(pair):
            if not isinstance(ri, core.Variable):
                raise TypeError('All r values must be Variables')
            if not isinstance(ki, core.Variable):
                raise TypeError('All knot values must be Variables')
```